sources/appendix-value-extraction.asciidoc

```diff
@@ -0,0 +1,437 @@
+	@Override
+	public void extractValues(List<?> originalValue, ValueReceiver receiver) {
+		int i = 0;
+		for ( Object object : originalValue ) {
+			receiver.indexedValue( object, "<collection element>", i );
+			i++;
+		}
+	}
+}
+----
+
+The right callback methods (`indexedValue`, `keyedValue` etc.) must be called in order to allow for proper construction of the property path as per the rules layed out in the BV 1.1 spec.
+
+If a non-null value is passed for `nodeName`, a path node of type `TYPE_PARAMETER` will be appended to the property path. That's desirable for collection types for instance. If null is passed, no node will be appended, resulting in the same path as if the constraint had been given on the element itself instead of a type parameter. That's desirable for pure "wrapper types" such as `Optional`.
+
+In some cases, an extractor applies to the entire extracted type itself (e.g. for arrays). In this case the `@ExtractedValue` annotation is to be given on the extracted type itself:
+
+[source,java]
+----
+class ObjectArrayValueExtractor implements ValueExtractor<@ExtractedValue Object[]> {
+
+	@Override
+	public void extractValues(Object[] originalValue, ValueReceiver receiver) {
+		int i = 0;
+		for ( Object object : originalValue ) {
+			receiver.indexedValue( object, "<collection element>", i );
+			i++;
+		}
+	}
+}
+----
+
+[NOTE]
+.Motivation for callback-style API
+====
+Instead of returning the extracted values from the method call, implementations of `ValueExtractor` pass the extracted values to the given receiver callback. This helps to avoid object allocations and allows to handle the case of a single extracted value (`Optional<T>`) and multiple extracted values (`List<T>`) in a uniform fashion.
+====
+
+==== Default extractors
+
+Compatible implementations provide extractors for the following types out of the box:
+
+* arrays of objects and all primitive data types
+* `java.util.Iterable` (when `@Valid` is given on the iterable element itself, the element and all its entries will be validated; this is to grant backwards compatability with BV 1.1; when `@Valid` is given on the type parameter of an iterable element, all the entries will be validated). A node of type `TYPE_PARAMETER` and with the name "<collection element>" will be added to the path when validating a type parameter constraint.
+* `java.util.Map` (when `@Valid` is given on the map element itself, the element and all its values will be validated; this is to grant backwards compatability with BV 1.1; when `@Valid` is given on the key type parameter, only the keys will be validated; when `@Valid` is given on the value type parameter, only the values will be validated. A node of type `TYPE_PARAMETER` and with the name "<map value>" will be added to the path when validating a type parameter constraint on the map value. A node of type `TYPE_PARAMETER` and with the name "<map key>" will be added to the path when validating a type parameter constraint on the map key.
```

Comment:
I'm not concerned too much about the node generated by the row extractor or the column extractor, I imagine these will do something like `valuereceiver.value(row, "<table row>")`

 I'm concerned about the `Table` value extractor because in essence you want the row and column to be part of the path to identify the value. Just like our node for Map value has `getKey()`. While pondering the problem I saw a few initial approaches:
- make nodes generified and offer a way to get to a generic `getKey` (in the general sense of key).
- keep `getKey` as is and return `Object`

In both case, the key object would need to be a composite key if you will

```
class RowAndColumnKey {
    Object getRow();
    Object getColumn();
    String toString(); //cool for the string reprensetation
}